### ISSUE TITLE
min() function not supported in sizes

### ISSUE THREAD

**Issue by foolip** [validator/validator#1585]
Code to reproduce error:

```html
<!DOCTYPE html>
<html lang="en">
<head>
<title>Test</title>
</head>
<body>
<img src="cat-800w.webp"
     srcset="cat-800w.webp 800w,
             cat-1600w.webp 1600w,
             cat-2400w.webp 2400w"
     sizes="min(100vw - 40px, 800px)"
     width="800" height="600" alt="cat">
</body>
</html>
```

(The idea is that the page has a 20px margin/padding to the left and right

Result:

> Error: Bad value `min(100vw - 40px, 800px)` for attribute `sizes` on element `img`: Bad media condition: Parse Error at `min(100vw - 40px,`.

The message suggests that the problem is parsing the (omitted) condition, but it seems like the problem is really that the validator doesn't understand the `min()` function.

Things that do pass validation:

- `sizes="(max-width: 840px) calc(100vw - 40px), 800px"` (equivalent but longer + less readable)
- `sizes="calc(100vw - 40px)"` (showing the problem isn't omitting the media condition)

I've confirmed in testing that Chrome and Firefox do understand `sizes="min(100vw - 40px, 800px)"`, so I think the validator shouldn't complain about this usage.

https://bugs.webkit.org/show_bug.cgi?id=184027 is the WebKit bug about supporting this.

cc @zcorpan

**Comment by sideshowbarker** [comment]
I think this is an upstream issue with the CSS-checker backend, so I’m pinging @ylafon to find out if maybe it’s a known issue, and maybe a GitHub issue has already been opened for it in the CSS checker’s issue tracker.

**Comment by ylafon** [comment]
`tst { width:min(100vw - 40px, 800px) }` validates, so it depends what is sent to the validator when extracted from the `sizes` attribute, is it supposed to be a value or a media selector?

**Comment by ylafon** [comment]
FWIW ```@media (width:min(100vw - 40px, 800px)) {
tst { color: green }
}``` also validates

**Comment by sideshowbarker** [comment]
> I think this is an upstream issue with the CSS-checker backend,

Actually, after looking more, I see now that we’re not using the CSS-checker backend to check `srcset` values, but instead using our own custom code at https://github.com/validator/validator/blob/main/src/nu/validator/datatype/ImageCandidateStrings.java

So this issue doesn’t actually relate to the CSS checker at all, it seems

**Comment by zcorpan** [comment]
`min()` goes in `sizes` not `srcset`, so I guess https://github.com/validator/validator/blob/main/src/nu/validator/datatype/SourceSizeList.java#L250 is the right place. Per https://github.com/validator/validator/blob/main/src/nu/validator/datatype/SourceSizeList.java#L98 it looks like it should be supported already?

Just `sizes="min(50px)"` validates but `sizes="min(50px, 20px)"` does not, so it seems like the it may be missing conversion to component values.

https://html.spec.whatwg.org/#parse-a-sizes-attribute 
->
https://drafts.csswg.org/css-syntax/#parse-a-comma-separated-list-of-component-values 
->
https://drafts.csswg.org/css-syntax/#consume-a-component-value
->
https://drafts.csswg.org/css-syntax/#consume-a-function
->
`min(50px, 20px)` is one component value.

**Comment by jfrech** [comment]
Toying around a bit, the issue to me seems to be detecting dyadic CSS functions ([minimal example](https://validator.w3.org/nu/?doc=data%3Atext%2Fhtml%2C%3C%21doctype+html%3E%3Chtml+lang%3D%22en%22%3E%3Chead%3E%3Cmeta+charset%3D%22utf-8%22%3E%3Ctitle%3E0%3C%2Ftitle%3E%3C%2Fhead%3E%3Cbody%3E%3Cpicture%3E%3Csource+sizes%3D%22min%280%2C0%29%22+srcset%3D%220+1w%22%3E%3Cimg+src%3D%220%22+alt%3D%220%22%3E%3C%2Fpicture%3E%3C%2Fbody%3E) (fixed as of 2026-03-22, see below for the once-present error)).

I would very much welcome the validator accept two-valued `min`.

<hr/>

Nu v22.3.8 and Nu v26.3.20 on the minimal example above:

<img width="1790" height="897" alt="Image" src="https://github.com/user-attachments/assets/ae12c482-c651-43ea-b941-4e0bb4f3d999" />
<img width="1790" height="897" alt="Image" src="https://github.com/user-attachments/assets/e613062b-7eb6-46f7-b518-b0964dd98747" />

**Comment by sideshowbarker** [comment]
I pushed https://github.com/validator/validator/commit/4f9c46647ff570a0f9fa8eaca5bf35396f03c3f2 with a change that fixes this. I think. At least it does for the case reported in issue description — and for similar cases (other functions).

There still may be some other function syntax that we choke on. But if so, I think any could be handled in new, separate PRs.
